File: examples/05-ops-and-multitenancy/python/ops_and_multitenancy.py
#!/usr/bin/env python3
"""Jigyasa Example 05 — Ops & Multi-Tenancy."""

import subprocess, sys, os, time, json

# ── stub generation ──────────────────────────────────────────────────────────
PROTO_DIR = os.path.join(os.path.dirname(__file__), "..", "..", "..", "src", "main", "proto")
GEN_DIR = os.path.join(os.path.dirname(__file__), "_gen")
os.makedirs(GEN_DIR, exist_ok=True)
import grpc_tools
grpc_include = os.path.join(os.path.dirname(grpc_tools.__file__), "_proto")
import site
site_pkgs = [p for p in site.getsitepackages() if "site-packages" in p][0]
subprocess.run([sys.executable, "-m", "grpc_tools.protoc",
    f"--proto_path={PROTO_DIR}", f"--proto_path={grpc_include}", f"--proto_path={site_pkgs}",
    f"--python_out={GEN_DIR}", f"--grpc_python_out={GEN_DIR}", "dpSearch.proto"
], capture_output=True, text=True, check=True)
sys.path.insert(0, GEN_DIR)
import dpSearch_pb2 as pb
import dpSearch_pb2_grpc as pb_grpc

import grpc

COLLECTION = "ops_demo"
ADDR = "localhost:50051"

SCHEMA = json.dumps({"fields": [
    {"name": "id", "type": "STRING", "key": True, "filterable": True},
    {"name": "title", "type": "STRING", "searchable": True},
    {"name": "category", "type": "STRING", "filterable": True},
]})

ACME_DOCS = [
    {"id": "a1", "title": "Acme Cloud Platform",      "category": "software"},
    {"id": "a2", "title": "Acme Rocket Engine",        "category": "hardware"},
    {"id": "a3", "title": "Acme Deployment Toolkit",   "category": "software"},
]

GLOBEX_DOCS = [
    {"id": "g1", "title": "Globex Strategy Report",    "category": "consulting"},
    {"id": "g2", "title": "Globex Analytics Suite",    "category": "software"},
]


def banner(msg: str):
    print(f"\n{'=' * 60}\n  {msg}\n{'=' * 60}")


def run():
    channel = grpc.insecure_channel(ADDR)
    stub = pb_grpc.JigyasaDataPlaneServiceStub(channel)

    # ── 1. Collection lifecycle ──────────────────────────────────────────────
    banner("1. Collection Lifecycle")

    print("Creating collection …")
    try:
        stub.CreateCollection(pb.CreateCollectionRequest(
            collection=COLLECTION, indexSchema=SCHEMA))
        print(f"  ✓ Created '{COLLECTION}'")
    except grpc.RpcError as e:
        if "already exists" in str(e.details()):
            print(f"  ✓ '{COLLECTION}' already exists")
        else:
            raise

    resp = stub.ListCollections(pb.ListCollectionsRequest())
    print(f"  Collections after create: {list(resp.collections)}")

    print("Closing collection …")
    stub.CloseCollection(pb.CloseCollectionRequest(collection=COLLECTION))
    print(f"  ✓ Closed '{COLLECTION}'")

    print("Reopening collection (schema from index) …")
    stub.OpenCollection(pb.OpenCollectionRequest(collection=COLLECTION))
    print(f"  ✓ Reopened '{COLLECTION}'")

    resp = stub.ListCollections(pb.ListCollectionsRequest())
    print(f"  Collections after reopen: {list(resp.collections)}")

    # ── 2. Health monitoring ─────────────────────────────────────────────────
    banner("2. Health Monitoring")

    health = stub.Health(pb.HealthRequest())
    print(f"  Server status : {pb.HealthResponse.Status.Name(health.status)}")
    for c in health.collections:
        print(f"  Collection '{c.name}': writer_open={c.writer_open}, "
              f"searcher_available={c.searcher_available}, "
              f"doc_count={c.doc_count}, segments={c.segment_count}")

    # ── 3. Multi-tenancy ─────────────────────────────────────────────────────
    banner("3. Multi-Tenancy — Indexing")

    acme_items = [pb.IndexItem(document=json.dumps(d), tenant_id="acme") for d in ACME_DOCS]
    stub.Index(pb.IndexRequest(item=acme_items, collection=COLLECTION))
    print(f"  Indexed {len(ACME_DOCS)} docs for tenant 'acme'")

    globex_items = [pb.IndexItem(document=json.dumps(d), tenant_id="globex") for d in GLOBEX_DOCS]
    stub.Index(pb.IndexRequest(item=globex_items, collection=COLLECTION))
    print(f"  Indexed {len(GLOBEX_DOCS)} docs for tenant 'globex'")

    time.sleep(2)  # NRT refresh

    banner("3b. Multi-Tenancy — Tenant-Scoped Queries")

    for tenant in ("acme", "globex"):
        count_resp = stub.Count(pb.CountRequest(
            collection=COLLECTION, tenant_id=tenant))
        print(f"  Tenant '{tenant}' doc count: {count_resp.count}")

    query_resp = stub.Query(pb.QueryRequest(
        collection=COLLECTION,
        text_query="platform",
        tenant_id="acme",
        top_k=10,
        include_source=True,
    ))
    print(f"\n  Query 'platform' scoped to 'acme' → {len(query_resp.hits)} hit(s):")
    for h in query_resp.hits:
        doc = json.loads(h.source)
        print(f"    {doc['id']}: {doc['title']}")

    query_resp = stub.Query(pb.QueryRequest(
        collection=COLLECTION,
        text_query="platform",
        tenant_id="globex",
        top_k=10,
        include_source=True,
    ))
    print(f"\n  Query 'platform' scoped to 'globex' → {len(query_resp.hits)} hit(s) (expected 0)")

    # ── 4. Delete by query ───────────────────────────────────────────────────
    banner("4. Delete By Query")

    del_resp = stub.DeleteByQuery(pb.DeleteByQueryRequest(
        collection=COLLECTION,
        tenant_id="acme",
        filters=[pb.FilterClause(field="category", term_filter=pb.TermFilter(value="hardware"))],
    ))
    print(f"  Deleted (category='hardware' for tenant 'acme'), count={del_resp.deleted_count}")

    time.sleep(1)

    count_resp = stub.Count(pb.CountRequest(
        collection=COLLECTION, tenant_id="acme"))
    print(f"  Tenant 'acme' doc count after delete: {count_resp.count}")

    # ── 5. Force merge ───────────────────────────────────────────────────────
    banner("5. Force Merge")

    merge_resp = stub.ForceMerge(pb.ForceMergeRequest(
        collection=COLLECTION, max_segments=1))
    print(f"  Segments before: {merge_resp.segments_before}")
    print(f"  Segments after : {merge_resp.segments_after}")

    # ── Cleanup ──────────────────────────────────────────────────────────────
    banner("Cleanup")
    stub.CloseCollection(pb.CloseCollectionRequest(collection=COLLECTION))
    print(f"  ✓ Closed '{COLLECTION}'")

    print("\n✅ Done.")
    channel.close()


if __name__ == "__main__":
    run()

File: examples/05-ops-and-multitenancy/java/src/main/java/opsdemo/OpsAndMultitenancy.java
package opsdemo;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ObjectNode;
import io.grpc.ManagedChannel;
import io.grpc.ManagedChannelBuilder;
import io.grpc.StatusRuntimeException;
import com.jigyasa.dp.search.protocol.*;
import com.jigyasa.dp.search.protocol.JigyasaDataPlaneServiceGrpc;

import java.util.List;
import java.util.Map;
import java.util.concurrent.TimeUnit;

/**
 * Jigyasa Example 05 — Ops &amp; Multi-Tenancy (Java).
 */
public class OpsAndMultitenancy {

    private static final String COLLECTION = "ops_demo_java";
    private static final String ADDR = "localhost:50051";
    private static final ObjectMapper MAPPER = new ObjectMapper();

    private static final String SCHEMA = "{\"fields\":["
            + "{\"name\":\"id\",\"type\":\"STRING\",\"key\":true,\"filterable\":true},"
            + "{\"name\":\"title\",\"type\":\"STRING\",\"searchable\":true},"
            + "{\"name\":\"category\",\"type\":\"STRING\",\"filterable\":true}"
            + "]}";

    private static final List<Map<String, String>> ACME_DOCS = List.of(
            Map.of("id", "a1", "title", "Acme Cloud Platform",    "category", "software"),
            Map.of("id", "a2", "title", "Acme Rocket Engine",      "category", "hardware"),
            Map.of("id", "a3", "title", "Acme Deployment Toolkit", "category", "software")
    );

    private static final List<Map<String, String>> GLOBEX_DOCS = List.of(
            Map.of("id", "g1", "title", "Globex Strategy Report",  "category", "consulting"),
            Map.of("id", "g2", "title", "Globex Analytics Suite",  "category", "software")
    );

    public static void main(String[] args) throws Exception {
        ManagedChannel channel = ManagedChannelBuilder.forTarget(ADDR).usePlaintext().build();
        JigyasaDataPlaneServiceGrpc.JigyasaDataPlaneServiceBlockingStub stub =
                JigyasaDataPlaneServiceGrpc.newBlockingStub(channel);

        try {
            collectionLifecycle(stub);
            healthMonitoring(stub);
            multiTenancy(stub);
            deleteByQuery(stub);
            forceMerge(stub);
            cleanup(stub);
        } finally {
            channel.shutdownNow().awaitTermination(5, TimeUnit.SECONDS);
        }
        System.out.println("\n✅ Done.");
    }

    // ── 1. Collection Lifecycle ─────────────────────────────────────────────

    private static void collectionLifecycle(
            JigyasaDataPlaneServiceGrpc.JigyasaDataPlaneServiceBlockingStub stub) {
        banner("1. Collection Lifecycle");

        System.out.println("Creating collection …");
        try {
            stub.createCollection(CreateCollectionRequest.newBuilder()
                    .setCollection(COLLECTION).setIndexSchema(SCHEMA).build());
            System.out.printf("  ✓ Created '%s'%n", COLLECTION);
        } catch (StatusRuntimeException e) {
            if (e.getMessage().contains("already exists")) {
                System.out.printf("  ✓ '%s' already exists%n", COLLECTION);
            } else {
                throw e;
            }
        }

        ListCollectionsResponse listResp = stub.listCollections(
                ListCollectionsRequest.getDefaultInstance());
        System.out.printf("  Collections: %s%n", listResp.getCollectionsList());

        System.out.println("Closing collection …");
        stub.closeCollection(CloseCollectionRequest.newBuilder()
                .setCollection(COLLECTION).build());
        System.out.printf("  ✓ Closed '%s'%n", COLLECTION);

        System.out.println("Reopening collection (schema from index) …");
        stub.openCollection(OpenCollectionRequest.newBuilder()
                .setCollection(COLLECTION).build());
        System.out.printf("  ✓ Reopened '%s'%n", COLLECTION);

        listResp = stub.listCollections(ListCollectionsRequest.getDefaultInstance());
        System.out.printf("  Collections after reopen: %s%n", listResp.getCollectionsList());
    }

    // ── 2. Health Monitoring ────────────────────────────────────────────────

    private static void healthMonitoring(
            JigyasaDataPlaneServiceGrpc.JigyasaDataPlaneServiceBlockingStub stub) {
        banner("2. Health Monitoring");

        HealthResponse health = stub.health(HealthRequest.getDefaultInstance());
        System.out.printf("  Server status : %s%n", health.getStatus());
        for (CollectionHealth c : health.getCollectionsList()) {
            System.out.printf("  Collection '%s': writer_open=%b, searcher=%b, "
                            + "docs=%d, segments=%d%n",
                    c.getName(), c.getWriterOpen(), c.getSearcherAvailable(),
                    c.getDocCount(), c.getSegmentCount());
        }
    }

    // ── 3. Multi-Tenancy ────────────────────────────────────────────────────

    private static void multiTenancy(
            JigyasaDataPlaneServiceGrpc.JigyasaDataPlaneServiceBlockingStub stub)
            throws Exception {
        banner("3. Multi-Tenancy — Indexing");

        indexDocs(stub, ACME_DOCS, "acme");
        indexDocs(stub, GLOBEX_DOCS, "globex");

        Thread.sleep(1000); // NRT refresh

        banner("3b. Multi-Tenancy — Tenant-Scoped Queries");

        for (String tenant : List.of("acme", "globex")) {
            CountResponse count = stub.count(CountRequest.newBuilder()
                    .setCollection(COLLECTION).setTenantId(tenant).build());
            System.out.printf("  Tenant '%s' doc count: %d%n", tenant, count.getCount());
        }

        QueryResponse qr = stub.query(QueryRequest.newBuilder()
                .setCollection(COLLECTION).setTextQuery("platform")
                .setTenantId("acme").setTopK(10).setIncludeSource(true).build());
        System.out.printf("%n  Query 'platform' scoped to 'acme' → %d hit(s):%n",
                qr.getHitsCount());
        for (QueryHit h : qr.getHitsList()) {
            var doc = MAPPER.readTree(h.getSource());
            System.out.printf("    %s: %s%n",
                    doc.get("id").asText(), doc.get("title").asText());
        }

        qr = stub.query(QueryRequest.newBuilder()
                .setCollection(COLLECTION).setTextQuery("platform")
                .setTenantId("globex").setTopK(10).setIncludeSource(true).build());
        System.out.printf("%n  Query 'platform' scoped to 'globex' → %d hit(s) (expected 0)%n",
                qr.getHitsCount());
    }

    // ── 4. Delete By Query ──────────────────────────────────────────────────

    private static void deleteByQuery(
            JigyasaDataPlaneServiceGrpc.JigyasaDataPlaneServiceBlockingStub stub)
            throws Exception {
        banner("4. Delete By Query");

        DeleteByQueryResponse delResp = stub.deleteByQuery(DeleteByQueryRequest.newBuilder()
                .setCollection(COLLECTION)
                .setTenantId("acme")
                .addFilters(FilterClause.newBuilder()
                        .setField("category")
                        .setTermFilter(TermFilter.newBuilder().setValue("hardware")))
                .build());
        System.out.printf("  Deleted category='hardware' for tenant 'acme' (count=%d)%n",
                delResp.getDeletedCount());

        Thread.sleep(1000);

        CountResponse count = stub.count(CountRequest.newBuilder()
                .setCollection(COLLECTION).setTenantId("acme").build());
        System.out.printf("  Tenant 'acme' doc count after delete: %d%n", count.getCount());
    }

    // ── 5. Force Merge ──────────────────────────────────────────────────────

    private static void forceMerge(
            JigyasaDataPlaneServiceGrpc.JigyasaDataPlaneServiceBlockingStub stub) {
        banner("5. Force Merge");

        ForceMergeResponse mergeResp = stub.forceMerge(ForceMergeRequest.newBuilder()
                .setCollection(COLLECTION).setMaxSegments(1).build());
        System.out.printf("  Segments before: %d%n", mergeResp.getSegmentsBefore());
        System.out.printf("  Segments after : %d%n", mergeResp.getSegmentsAfter());
    }

    // ── Cleanup ─────────────────────────────────────────────────────────────

    private static void cleanup(
            JigyasaDataPlaneServiceGrpc.JigyasaDataPlaneServiceBlockingStub stub) {
        banner("Cleanup");
        stub.closeCollection(CloseCollectionRequest.newBuilder()
                .setCollection(COLLECTION).build());
        System.out.printf("  ✓ Closed '%s'%n", COLLECTION);
    }

    // ── Helpers ─────────────────────────────────────────────────────────────

    private static void indexDocs(
            JigyasaDataPlaneServiceGrpc.JigyasaDataPlaneServiceBlockingStub stub,
            List<Map<String, String>> docs, String tenant) throws Exception {
        IndexRequest.Builder req = IndexRequest.newBuilder().setCollection(COLLECTION);
        for (Map<String, String> doc : docs) {
            req.addItem(IndexItem.newBuilder()
                    .setDocument(MAPPER.writeValueAsString(doc))
                    .setTenantId(tenant)
                    .build());
        }
        stub.index(req.build());
        System.out.printf("  Indexed %d docs for tenant '%s'%n", docs.size(), tenant);
    }

    private static void banner(String msg) {
        System.out.println();
        System.out.println("=".repeat(60));
        System.out.printf("  %s%n", msg);
        System.out.println("=".repeat(60));
    }
}
